src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSwapAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function createBaseApproveTransaction(
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction
+) {
+  const { guaranteedPrice, sellTokenAddress, buyAmount, sellAmount, allowanceTarget, from } =
+    unvalidatedSwapTransaction
+  const amountType: string =
+    updatedField === Field.TO ? ('buyAmount' as const) : ('sellAmount' as const)
+
+  const amountToApprove =
+    amountType === 'buyAmount'
+      ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
+      : BigInt(sellAmount)
+
+  const data = encodeFunctionData({
+    abi: erc20.abi,
+    functionName: 'approve',
+    args: [allowanceTarget as Address, amountToApprove],
+  })
+
+  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
+    from: from as Address,
+    to: sellTokenAddress as Address,
+    data,
+    type: 'cip42',
+    amountToApprove,
+  }
+
+  return approveTx
+}
+
+function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
+  const { from, to, value, data, gas } = unvalidatedSwapTransaction
+
+  const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
+    from: from as Address,
+    to: to as Address,
+    value: BigInt(value ? value : 0),
+    data: data as Hex,
+    gas: BigInt(gas), // This isn't entirely accurate
+    type: 'cip42',
+  }
+
+  return swapTx
+}
+
+function getMaxGasCost(txs: TransactionRequestCIP42[]) {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return maxGasCost
+}
+
+async function prepareTransactions(
+  fromToken: TokenBalanceWithAddress,
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction,
+  price: string,
+  feeCurrencies: TokenBalance[]
+) {
+  const baseApproveTx = createBaseApproveTransaction(updatedField, unvalidatedSwapTransaction)
+  const baseSwapTx = createBaseSwapTransaction(unvalidatedSwapTransaction)
+
+  const maxGasCosts: Array<{ feeCurrency: TokenBalance; maxGasCostInDecimal: BigNumber }> = []
+  for (const feeCurrency of feeCurrencies) {
+    if (feeCurrency.balance.isLessThanOrEqualTo(0)) {
+      // No balance, try next fee currency
+      continue
+    }
+
+    const feeCurrencyAddress = !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+    const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+      publicClient.celo,
+      feeCurrencyAddress
+    )
+
+    const approveTx = {
+      ...baseApproveTx,
+      maxFeePerGas,
+      maxPriorityFeePerGas,
+      feeCurrency: feeCurrencyAddress,
+    }
+
+    // TODO maybe cache this? and add static padding when using non-native fee currency
```

Comment:
who's going to do this? is there a Linear issue tracking the work that can be linked to here? If not, how are we going to remember to do it?

Reply:
isn't the static padding part done in line 170?

Reply:
> who's going to do this? is there a Linear issue tracking the work that can be linked to here? If not, how are we going to remember to do it?

I am, this PR is just the beginning of more changes I'm doing.

Reply:
> isn't the static padding part done in line 170?

Line 170 is for the swap TX, this is for the approval.